wcコマンドを作るのプログラムを作成しました

a-kuroki-gs opened the pull request:
ls.rbは前回のlsコマンドを作る5と同様
wc.rbは新たに作成した

a-kuroki-gs committed:
wcコマンドを作るのプログラムを作成しました

ls.rbは前回のlsコマンドを作る5と同様
wc.rbは新たに作成した

a-kuroki-gs committed:
不要なコメントを削除

maedana commented:
ls.rbは別にコミットしなくてOKです。

maedana commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,124 @@
+  file_info[:words], max_length[:words] = count_words(file, max_length[:words])
+  file_info[:bytes], max_length[:bytes] = count_bytes(file, max_length[:bytes])
+
+  max_length =
```
```ruby
max_length = { lines: 0, words: 0, bytes: 0, path: 0 }
```
がすでに定義されてますよね。同一ローカル変数を意図が異なるもので上書き(HashだったものをIntegerにしている)してしまうのは可読性を下げるのでやめましょう。

a-kuroki-gs replied:
Integerだった`max_length`を`string_length`とし、
HashとIntegerの違いをわかりやすくしました。

maedana commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,124 @@
+opt.on('-c') { |v| params[:c] = v }
+opt.parse!(ARGV)
+
+file_info_array = []
```
配列は複数形で命名するようにしましょう。配列のような一般的な型で`_array`のように型名を後ろにつける命名はあまりしないです。

a-kuroki-gs replied:
`file_infos`に訂正しました。

maedana commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,124 @@
+    puts info[:path]
+  else
+    puts
+  end
```
`puts`と`puts nil`で結果が変わらないですし別に分岐する必要ないのではと思いました

a-kuroki-gs replied:
`nil`の表示についての理解が不足しておりました。
分岐せずに、`puts info[:path]`と変更しました。

maedana commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,124 @@
+else
+  ARGV.each do |file_path|
+    file_info = { lines: nil, words: nil, bytes: nil, path: nil }
+    file = File.read(file_path.to_s)
```
`file_path`はもともと文字列なので意味のないメソッド呼び出しをしてそうですね

a-kuroki-gs replied:
`to_s`メソッドを削除しました。

maedana commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,124 @@
+      7
+    else
+      0
+    end
```
これでは標準入力から大きな桁数のものが来たときにずれますねー。
![image](https://github.com/a-kuroki-gs/ruby-practices/assets/14562/13d1a31a-524f-4c83-af41-d34356eb115f)

a-kuroki-gs replied:
バイト数のみ大きな桁数、というのを想定できておりませんでした。
`0`ではなく`max_length.values.max - 2`(最大桁数 - 2)と変更しました。

maedana commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,124 @@
+
+require 'optparse'
+
+def count_lines(file, max_length)
```
メソッド名を複数形にしたのはどういう意図でしょうか? 配列を返すからですか?

それと、メソッド名からは行数を数えるのだなということしか伝わらないですね。この命名からmax_lengthも返すというのは違和感がありますがいかがでしょうか?

a-kuroki-gs replied:
行数は(ほとんどの場合)２行以上のようなニュアンスで、
行数＝lines という英語の認識違いをしておりました。

`max_length`の情報も抜けておりましたので、`get_line_info`と改名しました。

maedana commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,124 @@
+
+max_length = { lines: 0, words: 0, bytes: 0, path: 0 }
+
+if ARGV.empty?
```
ifのネストが深くてよみにくい感じがしませんか?

大きな括りで言うと
```ruby
if 標準入力のとき
  # file_info_arrayに必要な値を詰め込む ... 1
  # max_lengthを求める... 2
else # ファイルのとき
  # file_info_arrayに必要な値を詰め込む ... 3
  # max_lengthを求める ... 4
end

# file_info_array, max_lengthを元に表示する
```
なわけですよね。1,2,3,4をそれぞれメソッドにすることでスッキリするんじゃないかなと思います。必須ではないのですけど、可能ならばメソッド化の練習としてちょうどいい題材だと思うのでやってみて欲しいかなと思いました。いかがでしょうか。

a-kuroki-gs replied:
ここは自分でも迷った部分であり、
もっと簡単にできそうと感じつつも実現できませんでした。
ご指摘のおかげで全体像をつかみ、以前よりすっきりとまとめることができたかと思います。

`get_file_info`・`get_string_length_of_stdin`・`get_string_length_of_file`・`get_sum_info`メソッドを定義しました。

a-kuroki-gs committed:
wc.rbの修正をいくつか行いました

Integerだったmax_lengthをstring_lengthに変更
配列を複数形に変更
不要な分岐を削除(nill関連)
不要なメソッドの削除(to_sメソッド)
バイト数のmax_lengthの定義を修正
メソッド名の変更(get_line_info等)
ネストが深くなりすぎていたところにメソッドを作成(get_file_info等)

maedana commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,129 @@
+
+require 'optparse'
+
+def get_line_info(file, max_length)
```
ここ以外も同様ですが、Rubyにおいて `get_xxx`という命名はあまりしないかなーと思います。

```
attr_reader :hoge
```
ようにするといわゆる`getter`メソッドが定義されますけど、メソッド名は`hoge`ですよね。メソッド名の原則は動詞始まりですけど、名詞のメソッドの場合、暗黙的に`get_`が省略されているイメージかなって思います。

それと`_info`という命名が抽象的すぎるかなと思います。`_info`という命名からは何を返すのか想像がつきにくいですよね。今回そこまで求めませんが、僕だったら`行数取得のメソッド`と、`行数の最大値取得のメソッド`とかでわけるかなーと思います。可能な限り単一機能のメソッドの組み合わせでコードを書くとわかりやすいコードになるかなと思います。

maedana replied:
前回と似たような指摘になるのですが、max_lengthについて、メソッド内のスコープ限定とはいえ、別の場所でHashとして使われるパターンもあるので紛らわしいですね。`max_length`といわれると単なる数字をイメージするので、このメソッド内での使われ方が普通の使われ方だと思います。

Hashの方は複数のmax_lengthをまとめたものなので、Hashの方は命名変えたほうがわかりやすいんじゃないですかねー。

a-kuroki-gs replied:
- > Rubyにおいて `get_xxx`という命名はあまりしないかなーと思います。

`get`がついているメソッド名は`calculate` or `find`に変更しました。

- > `_info`という命名が抽象的すぎるかなと思います。
行数取得のメソッドと、行数の最大値取得のメソッドとかでわけるかなーと思います。

まず、`_info`の表記を変更しました。
(メソッド名は`data_and_max_length`・`sum_and_max_length`、変数名は`data_count`・配列名は`data_counts`に変更)
そしてメソッドを分割し、不要だと考えた行数・単語数・バイト数を求めるメソッドを削除しました。

- > Hashの方は複数のmax_lengthをまとめたものなので、Hashの方は命名変えたほうがわかりやすいんじゃないですかねー。

ハッシュの名前を`max_lengths_of_data_size`(メソッド定義内)と`string_lengths_of_data_size`(メソッド呼び出し時)に
変数名を`max_length`(メソッド定義内)と`string_length`(メソッド呼び出し時)に変更しました。

maedana commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,129 @@
+
+file_infos = []
+
+max_length = { lines: 0, words: 0, bytes: 0, path: 0 }
```
39行目から全部ですが、トップレベルで書くものは分散させないようにしましょう。このあといくつかメソッド定義が続いてまたトップレベルの記述がはじまりますよね。メソッド定義は一番上か一番下にまとめてしまうとよいかと思います。ここで定義された`params`等が凄く下の方で使うので読みにくくなってます。

a-kuroki-gs replied:
メソッド定義を１番上にまとめました。

maedana commented on 05.wc/ls.rb:
```diff
@@ -0,0 +1,202 @@
+      print_filenames(target_dir)
+    end
+  end
+end
```
前回明確に書かなかったのですけど、提出物に不要なので削除したほうがよいです。ノイズになってしまいますので。

a-kuroki-gs replied:
ls.rbを削除しました。

a-kuroki-gs committed:
メソッド定義を１番上にまとめました

a-kuroki-gs committed:
わかりにくいハッシュと変数の名前を変更しました

ハッシュと変数の名前が混ざっておりわかりにくかったため修正
ハッシュの名前をmax_lengths_of_data_size(メソッド定義内)とstring_lengths_of_data_size(メソッド呼び出し時)に
変数名をmax_length(メソッド定義内)とstring_length(メソッド呼び出し時)に統一し、変更

a-kuroki-gs committed:
ファイルの情報を求める部分でメソッドの削除・追加を行いました

まず最大数を求めるメソッドcalculate_max_lengthを定義
これによって、行数を求めるメソッドと最大行数を求めるメソッドを分割できる
分割した結果、行数、単語数、バイト数を求めるためのコードは少なく、メソッド化する必要がないと判断
以上の理由から、行数、単語数、バイト数をそれぞれ求めるメソッドを削除

a-kuroki-gs committed:
メソッド名、変数名を変更しました

メソッド名につけるべきでないgetをつけていたため変更
計算するメソッドにはcalculate、探すメソッドにはfindをつける
また、infoではわかりにくいためinfoがついているものを変更
メソッド名はdata_and_max_length・sum_and_max_lengthに変更
変数名はdata_count・配列名はdata_countsに変更

a-kuroki-gs committed:
提出物とは関係ないls.rbを削除しました
